File: Aufg5/src/observer.rb
require 'celluloid'

class Actor
  include Celluloid

  def initialize(name)
    @name = name
  end

  def onrecieve(message)
    if message.is_a? String
      puts message + @name
    end

  end

end


class Observer
  include Celluloid

  def initialize
    @observers = []
  end

  def add(observer)
    if !(@observers.include? observer)
      @observers.push(observer)
    end
  end

  def remove(observer)
    if @observers.include? observer
      @observers.delete(observer)
    end
  end

  def sentMessages(message)
    @observers.each { |obs|
      obs.onrecieve(message)
    }
  end

  def test
    print(@observers)
  end
end



Actor_1 = Actor.new("actor 1")
Actor_2 = Actor.new("actor 2")

Master = Observer.new()

Master.add(Actor_1)
Master.add(Actor_2)
Master.sentMessages("Hello ")

Master.remove Actor_1
Master.sentMessages("Hello again ")

File: Aufg5/src/Observer.py
import pykka


class Actor(pykka.ThreadingActor):
    def __init__(self, name, my_arg=None):
        self.name = name
        super(Actor, self).__init__()

    def on_receive(self, message):
        if isinstance(message, dict):
            for value in message.values():
                print(value)
                print(self.name)


class Master(pykka.ThreadingActor):

    def __init__(self):
        self.observers = []
        super(Master, self).__init__()

    def remove(self, actorRef):
        if actorRef in self.observers:
            self.observers.remove(actorRef)

    def add(self, actorRef):
        if actorRef not in  self.observers:
            self.observers.append(actorRef)

    def sentMessages(self, message):
        for observer in self.observers:
            observer.tell(message)


def main():
    actor1_ref = Actor.start("actor 1")
    actor2_ref = Actor.start("actor 2")
    master = Master()
    master.start()

    master.add(actor1_ref)
    master.add(actor2_ref)
    master.sentMessages({"test": "hallo"})                             #actor.tell only allows dictionaries as parameter

    master.remove(actor1_ref)
    master.sentMessages({"test": "Test"})

    actor1_ref.stop(block=False)
    actor2_ref.stop(block=False)
    master.stop()


if __name__ == "__main__":
    main()
